xarray/core/dataset.py

```diff
@@ -2366,6 +2339,48 @@ def unstack(self, dim):
 
         return self._replace_vars_and_dims(variables, coord_names)
 
+    def unstack(self, dim=None):
+        """
+        Unstack existing dimensions corresponding to MultiIndexes into
+        multiple new dimensions.
+
+        New dimensions will be added at the end.
+
+        Parameters
+        ----------
+        dim : str or sequence of str, optional
+            Dimension(s) over which to unstack. By default unstacks all
+            MultiIndexes.
+
+        Returns
+        -------
+        unstacked : Dataset
+            Dataset with unstacked data.
+
+        See also
+        --------
+        Dataset.stack
+        """
+        if isinstance(dim, basestring):
+            dims = [dim]
+        elif dim is None:
+            dims = self.dims
+        else:
+            dims = dim
+
+        missing_dims = [d for d in dims if d not in self.dims]
+        if missing_dims:
+            raise ValueError('Dataset does not contain the dimensions: %s'
+                             % missing_dims)
```

Comment:
Looking at the implementation here, I see that it no longer raises an error when trying to unstack a dimension that does not use a MultiIndex. I'm not sure this is a desirable change -- on the one hand, it's nice to make unstacking an idempotent operation, but on the other hand, this might hide errors.

Instead, we could default `dims` to only include the entries from `self.dims` with a corresponding MultiIndex, and still raise an error when trying to unstack a MultiIndex.

Reply:
Right that was how I interpreted @fujiisoup's comment:

>I think that we can allow to unstack an object without MultiIndex, which just returns as is.
It would be useful if users want to remove any MultiIndexes from an object.

It seemed fine to me to have unstack just be a no-op if there are no MultiIndexes. But I can change it back if that would be better. What I was doing before was throwing an error if the user explicitly tries to unstack a non-MultiIndex, and otherwise just unstacking along all MultiIndexes unless there are none and throwing an error then.

Reply:
Ok so I made it so that if the user *explicitly tries to unstack a non-MultiIndex* they get an error. Otherwise it passes. So this is still valid `xr.Dataset().unstack()`.

Reply:
I was thinking something like:
```python
if dims is None:
    dims = [d for d in self.dims if isinstance(self.get_index(d), pd.MultiIndex)]
else:
    dims = [dim] if isinstance(dim, basestring) else dim
    # check for invalid dims

# call _unstack_once() repeatedly to unstack
```